opensea-wyvern-exchange-v1/src/modules/airstack/nft.ts

```diff
@@ -1,340 +1,463 @@
+      transaction.from = sellerAccount.id;
+      transaction.hash = txHash;
+
+      // Save Everything
+      buyerAccount.save();
+      sellerAccount.save();
+
+      sellAggregatedEntityAccount.save();
+      buyAggregatedEntityAccount.save();
+
+      contract.save();
+
+      saleStat.save();
+
+      transaction.save();
+
+      sellDailyAggregateEntityStats.save();
+      buyDailyAggregateEntityStats.save();
+
+      sellActionDailyAggregatedEntity.save();
+      buyActionDailyAggregatedEntity.save();
     }
-
-    export function getNFTSaleTransactionId(txHash: string, contractAddress: string, nftId: BigInt): string {
-
-        return dataSource.network()
-            .concat("-")
-            .concat(txHash)
-            .concat("-")
-            .concat(contractAddress)
-            .concat(nftId.toString());
+  }
+
+  export function getDailyAggregatedEntityId(
+    contractAddress: string,
+    timestamp: BigInt,
+    protocolActionType: string
+  ): string {
+    return dataSource
+      .network()
+      .concat("-")
+      .concat(contractAddress)
+      .concat("-")
+      .concat(protocolActionType.toString())
+      .concat("-")
+      .concat(getDaysSinceEpoch(timestamp.toI32()));
+  }
+
+  export function getNFTSaleTransactionId(
+    txHash: string,
+    contractAddress: string,
+    nftId: BigInt
+  ): string {
+    return dataSource
+      .network()
+      .concat("-")
+      .concat(txHash)
+      .concat("-")
+      .concat(contractAddress)
+      .concat(nftId.toString());
+  }
+
+  export function getAirNFTSaleStatsId(
+    contractAddress: string,
+    dailyAggregatedEntityId: string
+  ): string {
+    return dataSource
+      .network()
+      .concat("-")
+      .concat(dailyAggregatedEntityId)
+      .concat("-")
+      .concat(contractAddress);
+  }
+
+  export function getDailyAggregatedAccountId(
+    dailyAggregatedEntityId: string,
+    accountId: string
+  ): string {
+    return dailyAggregatedEntityId.concat("-").concat(accountId);
+  }
+
+  export function getAirDailyAggregateEntityStatsId(
+    dailyAggregatedEntityId: string
+  ): string {
+    return dailyAggregatedEntityId.concat("-").concat("stats");
+  }
+
+  export function getOrCreateAirContract(
+    contractAddress: Address
+  ): AirContract {
+    let entity = AirContract.load(contractAddress.toHexString());
+
+    if (entity == null) {
+      entity = new AirContract(contractAddress.toHexString());
+      entity.address = contractAddress.toHexString();
     }
-
-    export function getAirNFTSaleStatsId(contractAddress: string, dailyAggregatedEntityId: string): string {
-
-        return dataSource.network()
-            .concat("-")
-            .concat(dailyAggregatedEntityId)
-            .concat("-")
-            .concat(contractAddress)
-
+    return entity as AirContract;
+  }
+
+  export function getOrCreateAirDailyAggregateEntity(
+    id: string,
+    timestamp: BigInt,
+    protocolActionType: string,
+    blockHeight: BigInt
```

Comment:
May be this was not required, the blockheight could have be initialized as zero like others. The actual value is assigned at the place where this function is called.